Add long/float ColumnSelectorStrategy implementations

jon-wei opened the pull request:
This PR adds long and float implementations of the query processing ColumnSelectorStrategy interface introduced in https://github.com/druid-io/druid/pull/3570.

This allows users to include non-String columns as dimensions within queries by specifying a `typeHints` map in the query context. This is described in more detail in the doc changes within the PR.

For benchmarking, the patch also adds a set of "simple" schemas to the benchmark schemas, which has a single dimension "dimSequential" and a row count aggregator, where "dimSequential" contains an ascending sequence of integer values, encoded as String/Long/Float. Results for the benchmarks are located at:

https://gist.github.com/jon-wei/44a2b8ab97678710a9ef5b24aa107dec

---

NOTE: 
- When applying extraction functions to a float/long, the resulting DimensionSelector (exFns can only output Strings currently) constructs an on-the-fly dictionary without limits on memory usage, this may be problematic if an extraction function's set of output values has high cardinality. This is needed since query engines rely on the existence of a dictionary when handling Strings, I haven't thought of a better way to handle this yet.
---
NOTE: 
- Based on GroupByTypeInterfaceBenchmark, there is some virtual method overhead seen in GroupBy V2, like the concern discussed in this thread: https://github.com/druid-io/druid/pull/3797#discussion_r93735214.

```
Benchmark                                                                 (defaultStrategy)  (initialBuckets)  (numProcessingThreads)  (numSegments)  (queryGranularity)  (rowsPerSegment)  Mode  Cnt      Score     Error  Units
GroupByTypeInterfaceBenchmark.querySingleQueryableIndexFloatOnly                         v2                -1                       4              4                 all            250000  avgt   50  17895.755 ± 136.289  us/op
GroupByTypeInterfaceBenchmark.querySingleQueryableIndexFloatThenLong                     v2                -1                       4              4                 all            250000  avgt   50  35926.555 ± 110.157  us/op
GroupByTypeInterfaceBenchmark.querySingleQueryableIndexFloatThenString                   v2                -1                       4              4                 all            250000  avgt   50  44480.465 ± 398.283  us/op
GroupByTypeInterfaceBenchmark.querySingleQueryableIndexFloatTwice                        v2                -1                       4              4                 all            250000  avgt   50  35673.128 ±  40.837  us/op
GroupByTypeInterfaceBenchmark.querySingleQueryableIndexLongOnly                          v2                -1                       4              4                 all            250000  avgt   50  17920.575 ±  46.517  us/op
GroupByTypeInterfaceBenchmark.querySingleQueryableIndexLongThenString                    v2                -1                       4              4                 all            250000  avgt   50  46077.899 ± 470.782  us/op
GroupByTypeInterfaceBenchmark.querySingleQueryableIndexLongThenFloat                     v2                -1                       4              4                 all            250000  avgt   50  35986.157 ±  69.486  us/op
GroupByTypeInterfaceBenchmark.querySingleQueryableIndexLongTwice                         v2                -1                       4              4                 all            250000  avgt   50  35938.898 ±  70.856  us/op
GroupByTypeInterfaceBenchmark.querySingleQueryableIndexNumericOnly                       v2                -1                       4              4                 all            250000  avgt   50  20979.666 ±  51.640  us/op
GroupByTypeInterfaceBenchmark.querySingleQueryableIndexNumericThenString                 v2                -1                       4              4                 all            250000  avgt   50  52961.866 ± 438.055  us/op
GroupByTypeInterfaceBenchmark.querySingleQueryableIndexStringOnly                        v2                -1                       4              4                 all            250000  avgt   50  22801.356 ± 238.549  us/op
GroupByTypeInterfaceBenchmark.querySingleQueryableIndexStringThenLong                    v2                -1                       4              4                 all            250000  avgt   50  44720.358 ± 454.617  us/op
GroupByTypeInterfaceBenchmark.querySingleQueryableIndexStringThenNumeric                 v2                -1                       4              4                 all            250000  avgt   50  51805.316 ± 408.565  us/op
GroupByTypeInterfaceBenchmark.querySingleQueryableIndexStringTwice                       v2                -1                       4              4                 all            250000  avgt   50  46402.020 ± 657.067  us/op
```

This set of benchmarks attempts determine the running time for a group by query against single string/long/float columns (monomorphic base case), and running times for sequences of queries that use more than one type implementation.

For example, StringThenLong runs the string query followed by the long query within the same benchmark.

There appears to be some overhead when multiple type implementations are called, e.g.:

LongOnly = 17920.575us
StringOnly = 22801.356us

which should sum to ~40000us for both.

StringThenLong however has a running time of 444720.358us, about 10% higher than expected.

The difference bigger for NumericOnly (query on both a long and float) and NumericThenString:

NumericOnly = 20979.666
StringOnly = 22801.356

Sum should be ~44000, but NumericThenString and StringThenNumeric are 52961.866 and 51805.316 respectively.

The difference between StringOnly and StringTwice is smaller, suggesting the overhead may be related to the polymorphism. However, the behavior is not seen for long/float only, e.g. LongTwice is very close to LongThenFloat and (LongOnly + FloatOnly)


---

Similar benchmarks for TopN don't show the same discrepancy:

```
Benchmark                                                            (numSegments)  (rowsPerSegment)  (schemaAndQuery)  (threshold)  Mode  Cnt      Score      Error  Units
TopNTypeInterfaceBenchmark.querySingleQueryableIndexFloatOnly                    1            250000           basic.A           10  avgt   50  41916.788 ±  590.935  us/op
TopNTypeInterfaceBenchmark.querySingleQueryableIndexFloatThenLong                1            250000           basic.A           10  avgt   50  84411.958 ±  446.999  us/op
TopNTypeInterfaceBenchmark.querySingleQueryableIndexFloatThenString              1            250000           basic.A           10  avgt   50  94284.029 ± 1395.548  us/op
TopNTypeInterfaceBenchmark.querySingleQueryableIndexFloatTwice                   1            250000           basic.A           10  avgt   50  85390.807 ±  934.008  us/op
TopNTypeInterfaceBenchmark.querySingleQueryableIndexLongOnly                     1            250000           basic.A           10  avgt   50  43835.461 ±  612.863  us/op
TopNTypeInterfaceBenchmark.querySingleQueryableIndexLongThenFloat                1            250000           basic.A           10  avgt   50  86043.197 ±  597.789  us/op
TopNTypeInterfaceBenchmark.querySingleQueryableIndexLongThenString               1            250000           basic.A           10  avgt   50  87899.020 ±  520.239  us/op
TopNTypeInterfaceBenchmark.querySingleQueryableIndexLongTwice                    1            250000           basic.A           10  avgt   50  86441.232 ±  635.296  us/op
TopNTypeInterfaceBenchmark.querySingleQueryableIndexStringOnly                   1            250000           basic.A           10  avgt   50  50487.063 ±  402.854  us/op
TopNTypeInterfaceBenchmark.querySingleQueryableIndexStringThenFloat              1            250000           basic.A           10  avgt   50  90925.010 ±  731.045  us/op
TopNTypeInterfaceBenchmark.querySingleQueryableIndexStringThenLong               1            250000           basic.A           10  avgt   50  91756.314 ±  741.951  us/op
TopNTypeInterfaceBenchmark.querySingleQueryableIndexStringTwice                  1            250000           basic.A           10  avgt   50  98454.496 ±  996.865  us/op
```

---

I am looking into the GroupBy performance discrepancy more and considering other possible approaches (e.g., some form of code generation), but opening this PR now to get some thoughts from the community on these aspects.

gianm commented:
> When applying extraction functions to a float/long, the resulting DimensionSelector (exFns can only output Strings currently) constructs an on-the-fly dictionary without limits on memory usage, this may be problematic if an extraction function's set of output values has high cardinality. This is needed since query engines rely on the existence of a dictionary when handling Strings, I haven't thought of a better way to handle this yet.

SingleScanTimeDimSelector has the same problem. IMO, the query engines should maintain String dictionaries on their own if the selector doesn't have one, which it can report by returning `CARDINALITY_UNKNOWN` from `getValueCardinality()`. The query engines are in the best position to manage memory in an appropriate way. I don't think you need to do this as part of this PR, but, I think it's the best way to go ultimately.

gianm commented:
> Based on GroupByTypeInterfaceBenchmark, there is some virtual method overhead seen in GroupBy V2, like the concern discussed in this thread: https://github.com/druid-io/druid/pull/3797#discussion_r93735214.

> StringThenLong however has a running time of 444720.358us, about 10% higher than expected.

Do you have benchmarks on grouping on high-cardinality numeric columns as strings vs. as longs? (same data, but stored/grouped as string vs. stored/grouped as long). I suspect that this will be a much bigger difference than 10%, and if so, given that:

- there is a lot of performance benefit on both topN and groupBy in grouping numeric data as longs instead of strings
- there is no measurable performance hit from polymorphism in topN
- there is measurable hit in groupBy, but groupBy is under a lot of flux anyway
- we are investigating ways to un-do the performance hit

I'm willing to say that the change is worth including.

leventov commented:
> > When applying extraction functions to a float/long, the resulting DimensionSelector (exFns can only output Strings currently) constructs an on-the-fly dictionary without limits on memory usage, this may be problematic if an extraction function's set of output values has high cardinality. This is needed since query engines rely on the existence of a dictionary when handling Strings, I haven't thought of a better way to handle this yet.
>
> SingleScanTimeDimSelector has the same problem. IMO, the query engines should maintain String dictionaries on their own if the selector doesn't have one, which it can report by returning CARDINALITY_UNKNOWN from getValueCardinality(). The query engines are in the best position to manage memory in an appropriate way. I don't think you need to do this as part of this PR, but, I think it's the best way to go ultimately.

Also `StringDimensionIndexer`'s DimensionSelector and `QueryableIndexStorageAdapter`'s `DimensionSelector`s have positive `getValueCardinality()` but don't have dictionaries when there is non-null `extractionFn`. When I worked on #3858 I first changed them to return `CARDINALITY_UNKNOWN` in this case, but a lot of tests started to fail. Having cardinality or not and having dictionary or not are different properties. This is handled in #3858 by adding `idLookup()` method to `DimensionSelector` interface.

leventov commented:
> I am looking into the GroupBy performance discrepancy more and considering other possible approaches (e.g., some form of code generation), but opening this PR now to get some thoughts from the community on these aspects.

I considered code generation when worked on #3798. Decided it's too difficult and changes the current source tree too much. So I suggest to resolve this by expanding #3798 to groupBy queries and other types of queries other than topN.

leventov commented on processing/src/main/java/io/druid/query/filter/SelectorDimFilter.java:
```diff
@@ -137,6 +139,34 @@ public boolean applyLong(long input)
+          initFloatValue();
+
+          if (valueAsFloat == null) {
+            return new DruidFloatPredicate()
```
Please extract a singleton

leventov commented on processing/src/main/java/io/druid/query/filter/SelectorDimFilter.java:
```diff
@@ -137,6 +139,34 @@ public boolean applyLong(long input)
+              @Override
+              public boolean applyFloat(float input)
+              {
+                return input == unboxedFloat;
```
Same NaN considerations as above

jon-wei replied:
Changed this to use float int bits

leventov commented on processing/src/main/java/io/druid/query/filter/SelectorDimFilter.java:
```diff
@@ -225,4 +255,18 @@ private void initLongValue()
+      if (floatsInitialized) {
+        return;
+      }
+      valueAsFloat = Floats.tryParse(value);
```
If caching resulting predicate directly instead of `valueAsFloat` (it's possible because predicates are immutable), `floatsInitialized` is not needed. Also, locking could be removed and benign race allowed.

jon-wei replied:
Cached the predicate instead

leventov commented on processing/src/main/java/io/druid/query/filter/SelectorDimFilter.java:
```diff
@@ -137,6 +139,34 @@ public boolean applyLong(long input)
             };
           }
         }
```
In the method which ends with this brace `Predicates.equalTo()` and singleton could be used.

jon-wei replied:
I changed the String predicate here to use `Predicates.equalTo()` and replaced the always false predicates with singletons

leventov commented on processing/src/main/java/io/druid/query/filter/SelectorDimFilter.java:
```diff
@@ -225,4 +255,18 @@ private void initLongValue()
       longsInitialized = true;
```
If caching resulting predicate directly instead of `valueAsLong` (it's possible because predicates are immutable), `longsInitialized` is not needed. Also, locking could be removed and benign race allowed.

jon-wei replied:
Cached the predicate instead

leventov commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/GroupByQueryEngineV2.java:
```diff
@@ -342,6 +350,92 @@ private void initializeGroupingKeyV2Dimension(
+        int keyBufferPosition, Object rowObj, int rowValIdx, ByteBuffer keyBuffer
+    )
+    {
+      // rows from a long column are always size 1
```
This comment is not clear to me

jon-wei replied:
I've reworded the comment to mention multivalue support, and replaced the constant with a named one

leventov commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/GroupByQueryEngineV2.java:
```diff
@@ -342,6 +350,92 @@ private void initializeGroupingKeyV2Dimension(
+        int keyBufferPosition, Object rowObj, int rowValIdx, ByteBuffer keyBuffer
+    )
+    {
+      // rows from a float column are always size 1
```
This comment is not clear to me

leventov commented on processing/src/main/java/io/druid/query/topn/BaseTopNAlgorithm.java:
```diff
@@ -74,7 +74,10 @@ public void run(
     boolean hasDimValSelector = (dimValSelector != null);
 
-    final int cardinality = params.getCardinality();
+    int cardinality = params.getCardinality();
```
Why this change should be a part of this PR? cardinality < 0 probably means that completely different approach should be taken to compute topN, setting just to `Integer.MAX_VALUE` may result in a lot of excessive work and allocations in `run()` method.

jon-wei replied:
I've changed BaseTopNAlgorithm to have a "with cardinality" path and "no cardinality" path, the "no cardinality" path doesn't use `DimValSelector` (current implementations need cardinality) and the multi-pass logic

jon-wei commented:
@leventov @gianm 

Thanks for the reviews so far, marking this WIP while I address the comments and get more benchmark results, will update this early next week

leventov commented on processing/src/main/java/io/druid/segment/LongWrappingDimensionSelector.java:
```diff
@@ -0,0 +1,112 @@
+import java.util.List;
+import java.util.Map;
+
+public class LongWrappingDimensionSelector implements DimensionSelector
```
Same comments as in `FloatWrappingDimensionSelector`

leventov commented on processing/src/main/java/io/druid/segment/filter/SpatialFilter.java:
```diff
@@ -91,6 +92,20 @@ public boolean applyLong(long input)
+          @Override
+          public DruidFloatPredicate makeFloatPredicate()
+          {
+            return new DruidFloatPredicate()
```
Please replace with a singleton

leventov replied:
Please replace with a singleton

jon-wei replied:
whoops, missed that the first time, thanks

leventov commented on processing/src/main/java/io/druid/query/filter/BoundDimFilter.java:
```diff
@@ -343,83 +341,197 @@ public boolean applyLong(long input)
+        }
+      }
+
+      private void initPredicate()
```
Actually I'm losing the point of lazy initialization here, as well several similar places below. Maybe it's simpler to compute build a predicate eagerly.

gianm commented on processing/src/main/java/io/druid/query/topn/types/StringTopNColumnSelectorStrategy.java:
```diff
@@ -58,15 +62,29 @@ public int getCardinality(DimensionSelector selector)
   {
-    final IndexedInts dimValues = selector.getRow();
+    if (selector.getValueCardinality() < 0) {
+      dimExtractionScanAndAggregateNoCardinality(selector, rowSelector, aggregatesStore, cursor, query);
```
Hm, I guess the purpose of this is to support extractionFns on longs/floats?

gianm replied:
This'll probably need some reconciling with #3858.

jon-wei commented on benchmarks/src/main/java/io/druid/benchmark/DimensionPredicateFilterBenchmark.java:
```diff
@@ -95,6 +96,19 @@ public boolean applyLong(long input)
+        @Override
+        public DruidFloatPredicate makeFloatPredicate()
+        {
+          return new DruidFloatPredicate()
```
Using singletons for these false predicates now, in DruidLongPredicate/DruidFloatPredicate

jon-wei commented on processing/src/main/java/io/druid/query/BaseQuery.java:
```diff
@@ -67,6 +70,61 @@
     return parseInt(query, "uncoveredIntervalsLimit", defaultValue);
   }
 
+  public static <T> Map<String, Object> getContextTypeHints(Query<T> query)
```
I've removed this function now, based on @gianm's comment about moving the type specification to DimensionSpec

jon-wei commented on processing/src/main/java/io/druid/query/BaseQuery.java:
```diff
@@ -67,6 +70,61 @@
+    return query.getContextValue(QueryContextKeys.TYPE_HINTS);
+  }
+
+  public static <T> Map<String, ValueType> getContextTypeHints(Query<T> query, List<DimensionSpec> dimensions)
```
I've removed this function now, based on @gianm's comment about moving the type specification to DimensionSpec

jon-wei commented on processing/src/main/java/io/druid/query/BaseQuery.java:
```diff
@@ -67,6 +70,61 @@
+    }
+  }
+
+  private static <T> Map<String, ValueType> convertTypeHintsValues(Map<String, Object> oldMap)
```
Function was removed

jon-wei commented on processing/src/main/java/io/druid/query/aggregation/cardinality/types/LongCardinalityAggregatorColumnSelectorStrategy.java:
```diff
@@ -0,0 +1,40 @@
+  @Override
+  public void hashValues(LongColumnSelector dimSelector, HyperLogLogCollector collector)
+  {
+    collector.add(CardinalityAggregator.hashFn.hashLong(dimSelector.get()).asBytes());
```
I had it convert to string because of the lack of `hashFloat`, but changed this to use int bits now, thanks

jon-wei commented on processing/src/main/java/io/druid/query/filter/BoundDimFilter.java:
```diff
@@ -317,22 +324,13 @@ public String toString()
-        initLongData();
+        initPredicate();
 
         if (matchesNothing) {
```
Moved the `matchesNothing` block out and changed the predicate assignment

jon-wei commented on processing/src/main/java/io/druid/query/filter/FloatValueMatcherColumnSelectorStrategy.java:
```diff
@@ -0,0 +1,56 @@
+      @Override
+      public boolean matches()
+      {
+        return selector.get() == matchVal;
```
Changed this to compare using float int bits

jon-wei commented on processing/src/main/java/io/druid/query/filter/InDimFilter.java:
```diff
@@ -258,19 +262,17 @@ public String toString()
+          if (predicate != null) {
             return;
           }
 
```
cool, changed this to use LongArrayList

jon-wei commented on processing/src/main/java/io/druid/query/filter/InDimFilter.java:
```diff
@@ -282,46 +284,109 @@ private void initLongValues()
-          if (useLongHash) {
-            longHashSet = new HashSet<Long>(longs);
+          if (longs.size() > LONG_HASHING_THRESHOLD) {
+            final HashSet<Long> longHashSet = new HashSet<>(longs);
```
Changed this to use `LongOpenHashSet`

jon-wei commented on processing/src/main/java/io/druid/query/filter/InDimFilter.java:
```diff
@@ -282,46 +284,109 @@ private void initLongValues()
+            return;
+          }
+
+          List<Float> floats = new ArrayList<>();
```
Changed this to use an `IntArrayList` and float int bits

jon-wei commented on processing/src/main/java/io/druid/query/filter/InDimFilter.java:
```diff
@@ -282,46 +284,109 @@ private void initLongValues()
+          }
+
+          if (floats.size() > FLOAT_HASHING_THRESHOLD) {
+            final HashSet<Float> floatHashSet = new HashSet<>(floats);
```
Changed this to use `IntOpenHashSet` and float int bits

jon-wei commented on processing/src/main/java/io/druid/query/filter/InDimFilter.java:
```diff
@@ -282,46 +284,109 @@ private void initLongValues()
+              }
+            };
+          } else {
+            final float[] floatArray;
```
Changed this to use float int bits and int[]

jon-wei commented on processing/src/main/java/io/druid/segment/DimensionHandlerUtils.java:
```diff
@@ -138,15 +141,21 @@ public static DimensionHandler getHandlerFromCapabilities(
+      capabilities = DEFAULT_STRING_CAPABILITIES;
+    }
+
+    // Currently, all extractionFns output Strings, so force the type if present
```
Changed comment to:

"Currently, all extractionFns output Strings, so the column will return String values via a DimensionSelector if an extractionFn is present."

jon-wei commented on processing/src/main/java/io/druid/query/topn/types/StringTopNColumnSelectorStrategy.java:
```diff
@@ -46,6 +46,10 @@ public int getCardinality(DimensionSelector selector)
     // Optimization possibly requires a reverse lookup from value to ID, which is
     // not possible when applying an extraction function
 
+    if (params.getCardinality() < 0) {
```
removed the check, no longer needed with separate code path for "cardinality unknown" case

jon-wei commented on processing/src/main/java/io/druid/segment/DimensionHandlerUtils.java:
```diff
@@ -172,10 +185,56 @@ private static ColumnValueSelector getColumnValueSelectorFromDimensionSpec(
+
+  private static final Pattern LONG_PAT = Pattern.compile("[-|+]?\\d+");
+
+  // TopNBinaryFn relies on this returning the same Long object when no conversion is needed
```
I've removed this comment, I moved the type conversions to where DimExtractionTopNAlgorithm adds results to the result builder, so returning the same object is no longer strictly necessary (still returns the same object if type is the same though)

jon-wei commented on processing/src/main/java/io/druid/segment/DimensionHandlerUtils.java:
```diff
@@ -172,10 +185,56 @@ private static ColumnValueSelector getColumnValueSelectorFromDimensionSpec(
+      return ((Number) valObj).longValue();
+    } else if (valObj instanceof String) {
+      try {
+        String s = ((String) valObj).replace(",", "");
```
I took this parsing code from `getLongMetric/getFloatMetric` in `MapBasedRow`initially, I kept the comma removal since these transform functions would be used on String columns with "numeric" data, and figured it would be useful to allow that format

I don't mind removing this if someone is against it though.

leventov replied:
It's puzzling. I think it's worth to understand why it was added in the first place, and if the reason is still valid, to add a comment on this line.

leventov replied:
Well, I suppose the purpose is to remove 1000s delimiter from numbers, e. g. `1,000,000` is `1000000`. But not in all locales, e. g. in Russian locale comma is a separator between integer part and fractional part in non-integer numbers.

IMO this is "too smart" and we shouldn't do this. Probably it should be discussed separately, not in this PR

jon-wei replied:
Hm, good point on the locales, I agree this is too smart, I've removed the comma removal and replaced these blocks with `GuavaUtils.tryParseLong` and `Floats.tryParse`

jon-wei commented on processing/src/main/java/io/druid/segment/DimensionHandlerUtils.java:
```diff
@@ -172,10 +185,56 @@ private static ColumnValueSelector getColumnValueSelectorFromDimensionSpec(
+    }
+  }
+
+  // TopNBinaryFn relies on this returning the same Float object when no conversion is needed
```
Removed the comment after changing where the transformation is applied

jon-wei commented on processing/src/main/java/io/druid/segment/FloatWrappingDimensionSelector.java:
```diff
@@ -0,0 +1,112 @@
+{
+  private final FloatColumnSelector delegate;
+  private final ExtractionFn exFn;
+  private final Map<String, Integer> valueToId = Maps.newHashMap();
```
Agree on having the dictionary management reside in the query engines and other users, I'm in favor of getting #3858 merged first so I'll hold off on changing this part for now

jon-wei commented on processing/src/main/java/io/druid/query/filter/InDimFilter.java:
```diff
@@ -52,11 +53,13 @@
   // determined through benchmark that binary search on long[] is faster than HashSet until ~16 elements
   // Hashing threshold is not applied to String for now, String still uses ImmutableSortedSet
   public static final int LONG_HASHING_THRESHOLD = 16;
+  public static final int FLOAT_HASHING_THRESHOLD = 16;
```
Collapsed into `NUMERIC_HASHING_THRESHOLD`

jon-wei commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java:
```diff
@@ -172,10 +182,27 @@
+            key[i] = Strings.nullToEmpty(value);
           } else {
-            value = (String) row.getRaw(query.getDimensions().get(i - dimStart).getOutputName());
+            // only V2 groupby supports non-String types
```
my intent was to clarify the condition there, since I thought it wasn't immediately apparent that the `isInputRaw == true` case applied to groupby V2

jon-wei commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java:
```diff
@@ -467,7 +499,7 @@ public RowBasedKey fromByteBuffer(ByteBuffer buffer, int position)
 
       for (int i = dimStart; i < key.length; i++) {
-        key[i] = dictionary.get(buffer.getInt(dimsPosition + (Ints.BYTES * (i - dimStart))));
+        serdeHelpers.get(i - dimStart).getFromByteBuffer(buffer, dimsPosition, i, key);
```
Added comment about writing to key[i]

jon-wei commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java:
```diff
@@ -612,5 +650,168 @@ private int addToDictionary(final String s)
+      return size;
+    }
+
+    private List<RowBasedKeySerdeHelper> getSerdeHelpers()
```
renamed to `makeSerdeHelpers`

jon-wei commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java:
```diff
@@ -612,5 +650,168 @@ private int addToDictionary(final String s)
+    {
+      int getKeyBufferValueSize();
+
+      boolean putToByteBuffer(RowBasedKey key, int idx);
```
Added javadoc comments for the RowBasedKeySerdeHelper interface

jon-wei commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java:
```diff
@@ -612,5 +650,168 @@ private int addToDictionary(final String s)
+      }
+
+      @Override
+      public boolean putToByteBuffer(RowBasedKey key, int idx)
```
Renamed to `putTokeyBuffer`

jon-wei commented on processing/src/main/java/io/druid/query/topn/TopNBinaryFn.java:
```diff
@@ -72,35 +113,38 @@ public TopNBinaryFn(
     TopNResultValue arg2Vals = arg2.getValue();
 
     for (DimensionAndMetricValueExtractor arg1Val : arg1Vals) {
-      retVals.put(arg1Val.getStringDimensionValue(dimension), arg1Val);
```
It wasn't used anymore, I've removed `getStringDimensionValue` now

jon-wei commented on processing/src/main/java/io/druid/query/topn/TopNBinaryFn.java:
```diff
@@ -136,4 +183,82 @@ public TopNBinaryFn(
+      case FLOAT:
+        return FLOAT_TRANSFORMER;
+      default:
+        throw new IAE("invalid type");
```
this is now in `TopNMapFn`, with the type included in error msg

jon-wei commented on processing/src/main/java/io/druid/query/topn/TopNBinaryFn.java:
```diff
@@ -136,4 +183,82 @@ public TopNBinaryFn(
+    }
+  }
+
+  private DimensionAndMetricValueExtractor getTransformedExtractor(DimensionAndMetricValueExtractor argVal)
```
I moved the transform functions to `TopNMapFn`, the transformation is now applied within `DimExtractionTopNAlgorithm` as it adds results to the result builder

jon-wei commented on processing/src/main/java/io/druid/query/topn/BaseTopNAlgorithm.java:
```diff
@@ -196,10 +199,6 @@ public BaseArrayProvider(
-
-      if (keepOnlyN < 0) {
-        throw new UnsupportedOperationException("Cannot operate on a dimension with no dictionary");
-      }
```
Restored this after splitting the base topN algorithm to have separate "with cardinality" and "unknown cardinality" paths

jon-wei commented on processing/src/main/java/io/druid/query/topn/types/FloatTopNColumnSelectorStrategy.java:
```diff
@@ -0,0 +1,67 @@
+  @Override
+  public int getCardinality(FloatColumnSelector selector)
+  {
+    return -1;
```
Added `TopNColumnSelectorStrategy.CARDINALITY_UNKNOWN`

jon-wei commented on processing/src/main/java/io/druid/query/BaseQuery.java:
```diff
@@ -67,6 +70,61 @@
+        getContextTypeHints(query)
+    );
+
+    if (dimensions != null) {
```
Moved this to DimensionSpec outputType

jon-wei commented:
Updated the patch, still waiting on benchmark results comparing performance of a groupBy on a String column with numeric data vs. its equivalent long and float versions

I've left the filter predicate locking/lazy initialization for now since we haven't reached a consensus on that yet, addressed the other comments in that area though

I'm in favor of getting #3858 merged before this PR, so I'll revisit the FloatWrapping/LongWrapping selectors after that

The latest update on this PR also changes the hash function used in GroupByV2, since benchmarks showed very poor hashing behavior on floats using the existing XOR and shift function. The benchmarks used are included.

jon-wei commented:
Added benchmark results for a simple one column group by, showing difference between using Strings vs. numeric types for numeric data

https://gist.github.com/jon-wei/44a2b8ab97678710a9ef5b24aa107dec

jon-wei commented:
Adjusting this now that #3858 is merged

jon-wei commented:
I've rebased on top of #3858, moved the dictionary management out of `Float/LongWrappingDimensionSelector` into a String strategy variant in `GroupByEngineV2`. 

It doesn't have memory management currently, let's have that be a follow-on PR?

leventov commented:
@jon-wei could we also merge #3823 before this PR?

jihoonson commented on processing/src/main/java/io/druid/query/topn/DimExtractionTopNAlgorithm.java:
```diff
@@ -108,19 +109,42 @@ protected void updateResults(
+          }
+
+          resultBuilder.addEntry(
+              entry.getKey(),
```
It looks to miss null handling.

jon-wei replied:
The entry won't be null, so I think this is fine. Adding a test with a TopN that uses DimExtraction algorithm with null key values

jihoonson replied:
Ok. Thanks.

jihoonson commented on processing/src/main/java/io/druid/query/topn/DimValHolder.java:
```diff
@@ -26,13 +26,13 @@
 {
   private final Object topNMetricVal;
-  private final String dimName;
+  private final Comparable dimName;
```
Does ```dimName``` represent a dimension value?

jon-wei replied:
It does, I think "dimValue" would probably be a better name but didn't feel like renaming it in this PR

jihoonson replied:
Ok.

jihoonson commented on processing/src/main/java/io/druid/query/topn/types/FloatTopNColumnSelectorStrategy.java:
```diff
@@ -0,0 +1,74 @@
+      TopNQuery query, TopNParams params, Capabilities capabilities
+  )
+  {
+    return null;
```
Is this method going to be implemented in any follow-on PRs? If so, please add some comments.

jon-wei replied:
Not that I can see for the near future, since the usages for that method are dependent on the column providing cardinality and integer encoded values, which don't apply to the long/floats

There is a note about that function returning null for numerics in the TopNColumnSelectorStrategy's javadoc comments

jihoonson replied:
Ok. Thanks.

jihoonson commented on processing/src/main/java/io/druid/query/topn/types/LongTopNColumnSelectorStrategy.java:
```diff
@@ -0,0 +1,74 @@
+      TopNQuery query, TopNParams params, Capabilities capabilities
+  )
+  {
+    return null;
```
Is this method going to be implemented in any follow-on PRs? If so, please add some comments.

jon-wei commented:
@leventov sure, sounds good, I'll be reviewing #3823 today

jon-wei commented on processing/src/main/java/io/druid/query/filter/BoundDimFilter.java:
```diff
@@ -317,109 +323,192 @@ public String toString()
+                if (hasUpperFloatBound) {
+                  upperComparing = Float.compare(upperFloatBound, input);
+                }
+                if (lowerStrict && upperStrict) {
```
moved that block into a static method

jon-wei commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/GroupByQueryEngineV2.java:
```diff
@@ -341,6 +349,94 @@ private void initializeGroupingKeyV2Dimension(
+        int keyBufferPosition, Object rowObj, int rowValIdx, ByteBuffer keyBuffer
+    )
+    {
+      keyBuffer.putLong(keyBufferPosition, (Long) rowObj);
```
hm, good catch, I've adjusted this function to just return false for longs/floats, with a comment on why (it only handles row values after idx 0 for multivalued rows, so single-valued numeric implementations just need to return false, the first value is handled by initGroupingKeyColumnValue())

jihoonson replied:
It looks ok for now.

jon-wei commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java:
```diff
@@ -366,7 +404,7 @@ public int compare(RowBasedKey key1, RowBasedKey key2)
     {
       for (int i = dimStart; i < key1.getKey().length; i++) {
-        final int cmp = ((String) key1.getKey()[i]).compareTo((String) key2.getKey()[i]);
+        final int cmp = ((Comparable) key1.getKey()[i]).compareTo(key2.getKey()[i]);
```
I tried changing the key within RowBasedKey to `Comparable[]` instead of `Object[]` and that results in some json deserialization errors, I think I'll leave it as is for this PR

jihoonson replied:
Ok.

leventov commented:
@jon-wei let us know when this is ready for review

jon-wei commented:
@leventov @gianm @jihoonson 

merged master, PR is ready for review again

jon-wei commented:
added another commit that changes FloatWrappingDimensionSelector and LongWrappingDimensionSelector to extend BaseSingleValueDimensionSelector from #3823

leventov commented on processing/src/main/java/io/druid/query/topn/types/FloatTopNColumnSelectorStrategy.java:
```diff
@@ -0,0 +1,74 @@
+      FloatColumnSelector selector,
+      Cursor cursor,
+      Aggregator[][] rowSelector,
+      Map<Comparable, Aggregator[]> aggregatesStore
```
Maybe parameterize the type of aggregatesStore in `TopNColumnSelectorStrategy`, so it could be `Int2ObjectMap` for `FloatTopNColumnSelectorStrategy` and `Long2ObjectMap` for `LongTopNColumnSelectorStrategy`?

jon-wei replied:
Parameterized this and added two new methods to the TopNColumnSelectorStrategy interface, one for getting the aggregate store and one for pulling results out of the aggregate store into the result builder

leventov commented on processing/src/main/java/io/druid/query/topn/types/StringTopNColumnSelectorStrategy.java:
```diff
@@ -58,15 +64,29 @@ public int getCardinality(DimensionSelector selector)
   {
-    final IndexedInts dimValues = selector.getRow();
+    if (selector.getValueCardinality() != DimensionSelector.CARDINALITY_UNKNOWN) {
+      dimExtractionScanAndAggregateWithCardinality(selector, rowSelector, aggregatesStore, cursor, query);
```
In `BaseTopNAlgorithm` you use a different naming: cardinalityKnown/Unknown

jon-wei replied:
Changed this to use Known/Unknown convention

leventov commented on processing/src/main/java/io/druid/query/topn/types/StringTopNColumnSelectorStrategy.java:
```diff
@@ -58,15 +64,29 @@ public int getCardinality(DimensionSelector selector)
+      Aggregator[][] rowSelector,
+      Map<Comparable, Aggregator[]> aggregatesStore,
+      Cursor cursor,
+      TopNQuery query
```
Suggested to make `query` the first param, because 1) it is the first in `dimExtractionScanAndAggregate()` 2) it is an "input" param

jon-wei replied:
Reordered parameters

leventov commented on processing/src/main/java/io/druid/query/topn/types/StringTopNColumnSelectorStrategy.java:
```diff
@@ -85,4 +105,27 @@ public void dimExtractionScanAndAggregate(
+      DimensionSelector selector,
+      Map<Comparable, Aggregator[]> aggregatesStore,
+      Cursor cursor,
+      TopNQuery query
```
Same

jon-wei replied:
Reordered parameters

leventov commented on processing/src/main/java/io/druid/segment/DimensionHandlerUtils.java:
```diff
@@ -172,10 +185,56 @@ private static ColumnValueSelector getColumnValueSelectorFromDimensionSpec(
+      return ((Number) valObj).floatValue();
+    } else if (valObj instanceof String) {
+      try {
+        return Float.valueOf(((String) valObj).replace(",", ""));
```
I can imagine it could become a source or bugs, because in some Locales comma is used as a separator between integer and fractional parts. So removing commas from `"1,0"` will silently produce `10.0f` instead of failing with exception.

jon-wei commented on benchmarks/src/main/java/io/druid/benchmark/DimensionPredicateFilterBenchmark.java:
```diff
@@ -86,14 +87,13 @@ public boolean apply(String input)
-              return false;
-            }
-          };
+          return DruidLongPredicate.FALSE_PREDICATE;
```
Renamed to ALWAYS_FALSE

jon-wei commented on benchmarks/src/main/java/io/druid/benchmark/DimensionPredicateFilterBenchmark.java:
```diff
@@ -86,14 +87,13 @@ public boolean apply(String input)
+        @Override
+        public DruidFloatPredicate makeFloatPredicate()
+        {
+          return DruidFloatPredicate.FALSE_PREDICATE;
```
Renamed to ALWAYS_FALSE

jon-wei commented on processing/src/main/java/io/druid/query/aggregation/cardinality/types/FloatCardinalityAggregatorColumnSelectorStrategy.java:
```diff
@@ -0,0 +1,40 @@
+import io.druid.query.aggregation.hyperloglog.HyperLogLogCollector;
+import io.druid.segment.FloatColumnSelector;
+
+public class FloatCardinalityAggregatorColumnSelectorStrategy implements CardinalityAggregatorColumnSelectorStrategy<FloatColumnSelector>
```
Split into 2 lines

jon-wei commented on processing/src/main/java/io/druid/query/aggregation/cardinality/types/LongCardinalityAggregatorColumnSelectorStrategy.java:
```diff
@@ -0,0 +1,40 @@
+import io.druid.query.aggregation.hyperloglog.HyperLogLogCollector;
+import io.druid.segment.LongColumnSelector;
+
+public class LongCardinalityAggregatorColumnSelectorStrategy implements CardinalityAggregatorColumnSelectorStrategy<LongColumnSelector>
```
Split into 2 lines

jon-wei commented on processing/src/main/java/io/druid/query/filter/BoundDimFilter.java:
```diff
@@ -312,114 +318,198 @@ public String toString()
-      private volatile boolean longsInitialized = false;
-
-      // Other fields are not volatile.
       private boolean matchesNothing;
```
Made this a local variable

jon-wei commented on processing/src/main/java/io/druid/query/filter/BoundDimFilter.java:
```diff
@@ -312,114 +318,198 @@ public String toString()
-          longsInitialized = true;
+          if (matchesNothing) {
+            predicate = DruidLongPredicate.FALSE_PREDICATE;
+          } else {
```
I've split this to create a separate predicate for each of the possible combinations

jon-wei commented on processing/src/main/java/io/druid/query/filter/BoundDimFilter.java:
```diff
@@ -312,114 +318,198 @@ public String toString()
     return new BoundLongPredicateSupplier();
   }
+
+  private Supplier<DruidFloatPredicate> makeFloatPredicateSupplier()
```
split this one as well

jon-wei commented on processing/src/main/java/io/druid/query/filter/InDimFilter.java:
```diff
@@ -258,70 +264,123 @@ public String toString()
+            return;
+          }
+
+          IntArrayList floatInts = new IntArrayList(values.size());
```
renamed to floatBits

jon-wei commented on processing/src/main/java/io/druid/query/filter/InDimFilter.java:
```diff
@@ -258,70 +264,123 @@ public String toString()
+          }
+
+          if (floatInts.size() > NUMERIC_HASHING_THRESHOLD) {
+            final IntOpenHashSet floatIntsHashSet = new IntOpenHashSet(floatInts);
```
renamed to floatBitsHashSet

jon-wei commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/GroupByQueryEngineV2.java:
```diff
@@ -341,6 +358,145 @@ private void initializeGroupingKeyV2Dimension(
     }
   }
 
+  // builds internal String<->Integer dictionary for DimensionSelectors that return false for nameLookupPossibleInAdvance()
```
Changed this to javadoc style

jon-wei commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/GroupByQueryEngineV2.java:
```diff
@@ -341,6 +358,145 @@ private void initializeGroupingKeyV2Dimension(
   }
 
+  // builds internal String<->Integer dictionary for DimensionSelectors that return false for nameLookupPossibleInAdvance()
+  private static class DictionaryBuildingStringGroupByColumnSelectorStrategy extends StringGroupByColumnSelectorStrategy
```
Moved these and GroupByColumnSelectorPlus into top level classes in a new 'column' package under 'epinephilinae'

jon-wei commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/GroupByQueryEngineV2.java:
```diff
@@ -341,6 +358,145 @@ private void initializeGroupingKeyV2Dimension(
+    {
+      final DimensionSelector dimSelector = (DimensionSelector) selector;
+      final IndexedInts row = dimSelector.getRow();
+      final IntArrayList newIds = new IntArrayList(row.size());
```
changed this to use int[] array

jon-wei commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/GroupByQueryEngineV2.java:
```diff
@@ -341,6 +358,145 @@ private void initializeGroupingKeyV2Dimension(
+      final DimensionSelector dimSelector = (DimensionSelector) selector;
+      final IndexedInts row = dimSelector.getRow();
+      final IntArrayList newIds = new IntArrayList(row.size());
+      for (int id : row) {
```
Changed this to use a loop with index

jon-wei commented on processing/src/main/java/io/druid/query/topn/BaseTopNAlgorithm.java:
```diff
@@ -105,6 +119,21 @@ public void run(
     }
   }
 
+  // This function currently handles TopNs on long and float columns, which do not provide cardinality or an ID lookup.
```
Converted to javadoc

jon-wei commented on processing/src/main/java/io/druid/query/topn/DimExtractionTopNAlgorithm.java:
```diff
@@ -108,19 +109,42 @@ protected void updateResults(
-        for (int i = 0; i < aggs.length; i++) {
-          vals[i] = aggs[i].get();
-        }
+    if (needsResultTypeConversion(params)) {
```
Changed this as a part of parameterizing the aggregatesStore, now this is handled within type specific implementations by checking for `valueTransformer == null`

jon-wei commented on processing/src/main/java/io/druid/query/topn/TopNMapFn.java:
```diff
@@ -20,15 +20,62 @@
+    @Override
+    public Object apply(Object input)
+    {
+      if (input == null) {
```
Added a convertObjectToString method in DimensionHandlerUtils and changed the 4 areas that you mentioned

jon-wei commented on processing/src/main/java/io/druid/query/topn/TopNParams.java:
```diff
@@ -43,10 +43,6 @@ protected TopNParams(
     this.cursor = cursor;
     this.cardinality = selectorPlus.getColumnSelectorStrategy().getCardinality(selectorPlus.getSelector());
     this.numValuesPerPass = numValuesPerPass;
-
```
This is removed since the TopNParams can now hold CARDINALITY_UNKNOWN for certain dimension selectors

jon-wei commented:
@leventov addressed your latest comments

leventov commented on processing/src/main/java/io/druid/query/topn/types/TopNColumnSelectorStrategy.java:
```diff
@@ -82,6 +95,22 @@ void dimExtractionScanAndAggregate(
+   */
+  void updateDimExtractionResults(
+      DimExtractionAggregateStoreType aggregatesStore,
+      Function<Object, Object> valueTransformer,
```
Please annotate `@Nullable`

jon-wei replied:
Added @Nullable

leventov commented on processing/src/main/java/io/druid/segment/DimensionHandlerUtils.java:
```diff
@@ -206,7 +207,10 @@ private static ColumnValueSelector getColumnValueSelectorFromDimensionSpec(
   }
 
-  private static final Pattern LONG_PAT = Pattern.compile("[-|+]?\\d+");
+  public static String convertObjectToString(Object valObj)
```
Suggested to use `java.util.Objects.toString(valObj, null)` instead. Didn't notice this method during the previous round of review

jon-wei replied:
Changed these to use Objects.toString and removed that method

leventov commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/column/DictionaryBuildingStringGroupByColumnSelectorStrategy.java:
```diff
@@ -0,0 +1,86 @@
+        newIds[i] = dictId;
+      }
+    }
+    valuess[columnIndex] = ArrayBasedIndexedInts.of(newIds, newIds.length);
```
Just `ArrayBasedIndexedInts.of(newIds)`

jon-wei replied:
removed length from parameters

leventov commented on processing/src/main/java/io/druid/query/filter/InDimFilter.java:
```diff
@@ -258,70 +264,123 @@ public String toString()
+              }
+            };
+          } else {
+            final int[] floatIntsArray = floatBits.toIntArray();
```
Suggested "floatBitsArray"

jon-wei replied:
renamed to floatBitsArray

leventov commented on processing/src/main/java/io/druid/query/filter/SelectorDimFilter.java:
```diff
@@ -212,17 +193,57 @@ public int hashCode()
+      if (valueAsFloat == null) {
+        floatPredicate = DruidFloatPredicate.ALWAYS_FALSE;
+      } else {
+        final int floatInts = Float.floatToIntBits(valueAsFloat);
```
Suggested "floatBits" or "valueBits"

jon-wei replied:
renamed to floatBits

jon-wei commented:
@leventov updated PR again

dclim commented on processing/src/main/java/io/druid/segment/FloatColumnSelector.java:
```diff
@@ -26,5 +26,8 @@
 public interface FloatColumnSelector extends ColumnValueSelector
 {
+  // Float columns do not currently support multi-value rows
+  public static int ROW_SIZE = 1;
```
This doesn't seem to be used

dclim commented on processing/src/main/java/io/druid/segment/LongColumnSelector.java:
```diff
@@ -23,5 +23,8 @@
 public interface LongColumnSelector extends ColumnValueSelector
 {
+  // Long columns do not currently support multi-value rows
+  public static int ROW_SIZE = 1;
```
This doesn't seem to be used

jon-wei commented:
@dclim Thanks for the review, addressed your comments

dclim commented:
👍

gianm commented:
Taking one more look since the last round of changes.

gianm commented on processing/src/test/java/io/druid/query/topn/TopNQueryRunnerTest.java:
```diff
@@ -3678,4 +3679,351 @@ public void testTopNWithExtractionFilterNoExistingValue()
   }
+
+  @Test
+  public void testFullOnTopNFloatColumn()
```
Similar to groupBy, please include more tests covering:

- Long columns other than `__time`
- DefaultDimensionSpec on a column that is actually string but output type is LONG or FLOAT
- DefaultDimensionSpec on a column that is actually long but output type is STRING

jon-wei replied:
Added tests for those cases

jon-wei commented on docs/content/querying/dimensionspecs.md:
````diff
@@ -15,9 +15,18 @@ The following JSON fields can be used in a query to operate on dimension values.
+}
 ```
 
+When specifying a DimensionSpec on a numeric column, the user should include the type of the column in the `outputType` field. This is necessary as it is possible for a column with given name to have different value types in different segments. Result merging may fail unless results of different type are converted to the type specified by `outputType`.
````
I reworded this area to say that results will be converted to the type specified by outputType before merging

jon-wei commented on processing/src/test/java/io/druid/query/groupby/GroupByQueryRunnerTest.java:
```diff
@@ -6933,12 +6875,230 @@ public void testGroupByCardinalityAggOnFloat()
     // CardinalityAggregator currently treats non-String columns as having all nulls, so cardinality is 1 for
```
Removed this

jon-wei commented on processing/src/test/java/io/druid/query/groupby/GroupByQueryRunnerTest.java:
```diff
@@ -6933,12 +6875,230 @@ public void testGroupByCardinalityAggOnFloat()
+  }
+
+  @Test
+  public void testGroupByLongColumn()
```
Added tests above

jon-wei commented on processing/src/test/java/io/druid/query/groupby/GroupByQueryRunnerTest.java:
```diff
@@ -6933,12 +6875,230 @@ public void testGroupByCardinalityAggOnFloat()
+
+    // Only GroupBy V2 supports grouping on non-Strings.
+    if (config.getDefaultStrategy().equals(GroupByStrategySelector.STRATEGY_V1)) {
+      expectedException.expect(UnsupportedOperationException.class);
```
Added expectMessages to these

jon-wei commented on processing/src/test/java/io/druid/query/groupby/GroupByQueryRunnerTest.java:
```diff
@@ -6933,12 +6875,230 @@ public void testGroupByCardinalityAggOnFloat()
+  public void testGroupByFloatColumn()
+  {
+    // Only GroupBy V2 supports grouping on non-Strings.
+    if (config.getDefaultStrategy().equals(GroupByStrategySelector.STRATEGY_V1)) {
```
Changed these to use expectedException

jon-wei commented on processing/src/main/java/io/druid/query/filter/FloatValueMatcherColumnSelectorStrategy.java:
```diff
@@ -0,0 +1,57 @@
+import io.druid.segment.DimensionHandlerUtils;
+import io.druid.segment.FloatColumnSelector;
+
+public class FloatValueMatcherColumnSelectorStrategy implements ValueMatcherColumnSelectorStrategy<FloatColumnSelector>
```
Added `FloatFilteringTest`

jon-wei commented on processing/src/main/java/io/druid/query/dimension/DefaultDimensionSpec.java:
```diff
@@ -65,19 +66,30 @@ public DimensionSpec apply(String input)
   private static final byte CACHE_TYPE_ID = 0x0;
   private final String dimension;
   private final String outputName;
+  private final ValueType outputType;
```
Added outputType to those methods

jon-wei commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/Groupers.java:
```diff
@@ -31,12 +31,28 @@ private Groupers()
+
+  /*
+   * This method was rewritten in Java from an intermediate step of the Murmur hash function in
+   * http://code.google.com/p/smhasher/source/browse/trunk/MurmurHash3.cpp, which contained the
```
Updated location

jon-wei commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java:
```diff
@@ -172,10 +179,27 @@
+            key[i] = Strings.nullToEmpty(value);
           } else {
-            value = (String) row.getRaw(query.getDimensions().get(i - dimStart).getOutputName());
+            // only GroupByV2 (isInputRaw == false) supports non-String types
```
Adjusted this area to create a value supplier for each column based on the row signature, when isInputRaw is true, and added a comment to the method on isInputRaw

jon-wei commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java:
```diff
@@ -172,10 +179,27 @@
+                valObj = DimensionHandlerUtils.convertObjectToFloat(valObj);
+                break;
+              default:
+                throw new IAE("invalid type");
```
Included type

jon-wei commented on processing/src/test/java/io/druid/query/groupby/GroupByQueryRunnerTest.java:
```diff
@@ -6933,12 +6875,230 @@ public void testGroupByCardinalityAggOnFloat()
+        .builder()
+        .setDataSource(QueryRunnerTestHelper.dataSource)
+        .setQuerySegmentSpec(QueryRunnerTestHelper.firstToThird)
+        .setDimensions(Lists.<DimensionSpec>newArrayList(new ExtractionDimensionSpec("__time", "time_alias", jsExtractionFn)))
```
Added tests for query on long and query on long __time

jon-wei commented on processing/src/test/java/io/druid/query/groupby/GroupByQueryRunnerTest.java:
```diff
@@ -6933,12 +6875,230 @@ public void testGroupByCardinalityAggOnFloat()
+        .builder()
+        .setDataSource(QueryRunnerTestHelper.dataSource)
+        .setQuerySegmentSpec(QueryRunnerTestHelper.firstToThird)
+        .setDimensions(Lists.<DimensionSpec>newArrayList(new DefaultDimensionSpec("index", "index_alias", ValueType.FLOAT)))
```
Added tests for these cases

jon-wei commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/GroupByQueryEngineV2.java:
```diff
@@ -174,173 +177,28 @@ public void close() throws IOException
         case STRING:
-          return new StringGroupByColumnSelectorStrategy();
+          DimensionSelector dimSelector = (DimensionSelector) selector;
+          if (dimSelector.nameLookupPossibleInAdvance()) {
```
Changed this to check cardinality instead

jon-wei commented on processing/src/main/java/io/druid/segment/LongWrappingDimensionSelector.java:
```diff
@@ -0,0 +1,41 @@
+
+  public LongWrappingDimensionSelector(LongColumnSelector selector, ExtractionFn extractionFn)
+  {
+    this.delegate = selector;
```
renamed class members

jon-wei commented on processing/src/main/java/io/druid/segment/LongWrappingDimensionSelector.java:
```diff
@@ -0,0 +1,41 @@
+  public LongWrappingDimensionSelector(LongColumnSelector selector, ExtractionFn extractionFn)
+  {
+    this.delegate = selector;
+    this.exFn = extractionFn;
```
Changed these to accept null exFn and just call String.valueOf()

jon-wei commented on processing/src/main/java/io/druid/segment/QueryableIndexStorageAdapter.java:
```diff
@@ -454,6 +454,14 @@ private DimensionSelector makeDimensionSelectorUndecorated(
                       }
 
+                      if (columnDesc.getCapabilities().getType() == ValueType.LONG) {
+                        return new LongWrappingDimensionSelector(makeLongColumnSelector(dimension), extractionFn);
```
The only case these would get created is if a query has a dimension spec on a base numeric column but with an extractionFn applied, so exFn was never null

jon-wei commented on processing/src/main/java/io/druid/query/topn/types/LongTopNColumnSelectorStrategy.java:
```diff
@@ -0,0 +1,112 @@
+  }
+
+  @Override
+  public void dimExtractionScanAndAggregate(
```
moved cursor iteration into the methods

jon-wei commented on processing/src/main/java/io/druid/segment/column/ValueType.java:
```diff
@@ -19,12 +19,23 @@
+  public static ValueType fromString(String name)
+  {
+    if (name == null) {
+      return ValueType.STRING;
```
Changed this to return null and adjusted users of valuetype

gianm commented on sql/src/main/java/io/druid/sql/calcite/table/RowSignature.java:
```diff
@@ -76,6 +76,21 @@ public static Builder builder()
     return new Builder();
   }
 
+  public static ValueType getValueTypeForSqlTypeName(SqlTypeName sqlTypeName)
```
Please move this to "Calcites" where most of the random utility functions should go.

gianm commented on sql/src/main/java/io/druid/sql/calcite/aggregation/ApproxCountDistinctSqlAggregator.java:
```diff
@@ -82,7 +85,13 @@ public Aggregation toDruidAggregation(
+        throw new ISE("Cannot translate sqlTypeName[%s] to Druid type for field[%s]", sqlTypeName, name);
+      }
+
+      final DimensionSpec dimensionSpec = rex.toDimensionSpec(rowSignature, null, null);
```
Better to pass `ValueType.STRING` in here explicitly and check for non-null outputType in toDimensionSpec.

gianm commented on sql/src/main/java/io/druid/sql/calcite/expression/RowExtraction.java:
```diff
@@ -124,7 +124,11 @@ public boolean isFilterable(final RowSignature rowSignature)
+  public DimensionSpec toDimensionSpec(
+      final RowSignature rowSignature,
+      final String outputName,
+      final ValueType outputType
```
This should have a `Preconditions.checkNotNull(outputType, "outputType")` to prevent bugs due to accidentally forgetting to pass in an outputType. Callers should pass in STRING if that's what they want.

gianm commented on processing/src/test/java/io/druid/query/groupby/GroupByQueryRunnerTest.java:
```diff
@@ -7571,4 +7561,79 @@ public void testGroupByNestedWithInnerQueryNumericsWithLongTime()
+    List<Row> expectedResults;
+
+    if (config.getDefaultStrategy().equals(GroupByStrategySelector.STRATEGY_V1)) {
+      expectedResults = Arrays.asList(
```
This should throw an exception. In general, groupBy v1 (which doesn't support outputType other than STRING) should throw an exception if outputType is anything other than STRING. The message should be something like "groupBy v1 only supports outputType STRING" which will give people a little push to use groupBy v2.

gianm commented on processing/src/main/java/io/druid/query/groupby/orderby/DefaultLimitSpec.java:
```diff
@@ -107,10 +108,22 @@ public int getLimit()
+        final StringComparator naturalComparator;
+        if (columnType == ValueType.STRING) {
+          naturalComparator = StringComparators.LEXICOGRAPHIC;
+        } else {
```
If outputType is `COMPLEX` or `null` this is not true. I know this is "impossible" but please check for it anyway.

gianm replied:
A good way to do this is have an `else { sortingNeeded = true; break; }`

gianm commented:
@jon-wei please also update documentations:

- sql.md (the "Unsupported features" section)
- dimensionspecs.md (add "outputType" to the extraction spec, and note where outputTypes are and are not supported out of the places that use dimension specs: topN, groupBy v1, groupBy v2, cardinality aggregator, search query, select query)

gianm commented on processing/src/main/java/io/druid/segment/QueryableIndexStorageAdapter.java:
```diff
@@ -455,10 +456,17 @@ private DimensionSelector makeDimensionSelectorUndecorated(
 
                       if (columnDesc.getCapabilities().getType() == ValueType.LONG) {
+                        // Filtered dimension specs are not supported on numerics
+                        if (dimensionSpec instanceof BaseFilteredDimensionSpec) {
```
Ugh, these `instanceof BaseFilteredDimensionSpec` checks everywhere are kind of gnarly. There's way too many of them for comfort. Is there something better we can do? Maybe making them work is a better option if it's not too hard… why don't they, again? They should be able to decorate a wrapped dimension selector just fine, right?

Is it just that we don't know if a DimensionSpec's "decorate" method is actually going to get called, because getColumnValueSelectorFromDimensionSpec might skip it by going and making something like a LongColumnSelector instead? If so, we could address that by adding a "mustDecorate" method to DimensionSpec, and if it mustDecorate, then we force creation of a DimensionSelector (rather than whatever type the underlying column is).

I would also accept alternate solutions that don't involve instanceof checks all over the place.
